[GTK][WPE] Enable debug fission also when -g is present in the flags

calvaris opened the pull request:
#### 6f2f2e5ccba4b850192847fed1cc70b456170cbf
<pre>
[GTK][WPE] Enable debug fission also when -g is present in the flags
<a href="https://bugs.webkit.org/show_bug.cgi?id=251685">https://bugs.webkit.org/show_bug.cgi?id=251685</a>

Reviewed by Adrian Perez de Castro.

* Source/cmake/OptionsCommon.cmake:

Canonical link: <a href="https://commits.webkit.org/259957@main">https://commits.webkit.org/259957@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c99501e30c34af79a1dbb16e9dcea539cbeef1c7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106535 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15551 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39344 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115720 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17046 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6779 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98727 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112303 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95922 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94828 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27566 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96107 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8780 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28918 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/95551 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/6678 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9323 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5490 "Failed to checkout and rebase branch from PR 9602") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30650 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14939 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48466 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104294 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6885 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10861 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25840 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/cmake/OptionsCommon.cmake:
```diff
@@ -154,7 +154,7 @@ if (USE_THIN_ARCHIVES)
 
 set(ENABLE_DEBUG_FISSION_DEFAULT OFF)
-if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+if (CMAKE_BUILD_TYPE STREQUAL "Debug" OR "${CMAKE_CXX_FLAGS_RELEASE}" MATCHES "-g")
```
Hi, I think this won't work if build type is RelWithDebInfo or MinSizeRel or something custom, or if -g is passed via `CMAKE_CXX_FLAGS` rather than `CMAKE_CXX_FLAGS_RELEASE`. What CMake does is take `CMAKE_CXX_FLAGS` (which will default to using the environment variable `CXXFLAGS` if not specified), then append `CMAKE_CXX_FLAGS_<CONFIG>` ([reference)](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_FLAGS_CONFIG.html). So to reliably detect if a flag is passed, I think you need to check both `${CMAKE_CXX_FLAGS}` and `${CMAKE_CXX_FLAGS_$<CONFIG>}`. Who knows if that actually works, though.

(Also wonder if this would break distro find-debuginfo scripts, but the best way to find out is to proceed. ;)

calvaris replied:
It works, I tested it, at least downstream.

calvaris replied:
Actually downstream we don't build debug so we only care about release, that's why I used that variable. What do you think I should do to make it more general then? (I have trouble to properly understand your comment)

mcatanzaro replied:
Try `${CMAKE_BUILD_TYPE} STREQUAL "Debug" OR "${CMAKE_BUILD_TYPE} STREQUAL "RelWithDebInfo" OR ${CMAKE_CXX_FLAGS} MATCHES "-g" OR ${CMAKE_CXX_FLAGS_$<CONFIG>} MATCHES "-g"` (I'm hoping I have the syntax right for that generator expression...)

mcatanzaro commented on Source/cmake/OptionsCommon.cmake:
```diff
@@ -154,7 +154,7 @@ if (USE_THIN_ARCHIVES)
 
 set(ENABLE_DEBUG_FISSION_DEFAULT OFF)
-if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+if (CMAKE_BUILD_TYPE STREQUAL "Debug" OR CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo" OR "${CMAKE_CXX_FLAGS}" MATCHES "-g" OR "${CMAKE_CXX_FLAGS_RELEASE}" MATCHES "-g")
```
In particular, this doesn't work if build type is MinSizeRel or something custom.

Thinking about this more, we should definitely test for "RelWithDebInfo" build type, but I wonder if it's really valuable to add -gsplit-dwarf in response to a manually-added -g, since anyone doing that manually could also manually add -gsplit-dwarf. No strong opinion... as long as it doesn't break distro debuginfo, it's fine.

calvaris replied:
Can you r+ then, please?

mcatanzaro replied:
Almost. If you test  `${CMAKE_CXX_FLAGS_RELEASE}`, then you should at least also test `${CMAKE_CXX_FLAGS_MINSIZEREL}` as well, since that is the other standard build type that doesn't normally have debuginfo. I would r+ that.

But there is a disadvantage: you might be testing the flags for a build type that's not actually in use. E.g. `${CMAKE_CXX_FLAGS_RELEASE}` might really contain -g, but the user didn't select the Release build type. So best would be to try testing `${CMAKE_CXX_FLAGS_$<CONFIG>}` instead of `${CMAKE_CXX_FLAGS_RELEASE}`. I'm not certain if that will work, but if it does, then you have a simpler test that will check the flags for the configuration that's actually in use and none other and will work even for custom build types. But if that doesn't work, then don't worry about it.

mcatanzaro replied:
Background info: `$<CONFIG>` is a generator expression that expands to the current config name (which is generally, but not always, a build type).

The standard build types are:

CMake Build Type | CMake CFLAGS
-- | --
Debug | -g
Release | -O3 -DNDEBUG
RelWithDebInfo | -O2 -g -DNDEBUG
MinSizeRel | -Os -DNDEBUG

(`RelWithDebInfo` is the build type that best corresponds to standard distro builds and sounds like what you want to use downstream, but some distros like Fedora instead use `Release` and just override the flags. It doesn't really matter.)

calvaris replied:
The problem is that I assume it is not going to expand to the proper capitalized thing that we need to re-expand `${CMAKE_CXX_FLAGS_$<CONFIG>}`, right?

aperezdc replied:
There is no need to use the `$<CONFIG>` generator expression, there is `${CMAKE_BUILD_TYPE}`, and CMake supports indirect expansion of variables, so things like this work:

```cmake
string(TOUPPER "${CMAKE_BUILD_TYPE}" BUILD_TYPE_IN_CAPS)
message("Current build type CXX flags: ${CMAKE_CXX_FLAGS_${BUILD_TYPE_IN_CAPS}}")
```

Then, you want to use [separate_arguments()](https://cmake.org/cmake/help/v3.12/command/separate_arguments.html) instead of plain string matching, then check whether any of the items in the resulting list matches `-g<number>`, `-gdwarf`, `-gdwarf-<version>`, or `-ggdb`. With a plain regexp matching on `-g`, you will be matching other options that tweak how debuginfo is produced but have nothing to do with enabling debuginfo, and also other non-debuginfo compiler options like `-gen-decls` or `-gtoggle`...

So let's not go down this rabbit hole, or we would need to reimplement the same logic for command line parsing that the compiler does.

Instead, let's remove the check on the build type, and *always* add `-gsplit-dwarf` to the flags when supported by the compiler and linker. Why? Because this _modifies_ how debuginfo is generated, but very importantly does _not_ enable debuginfo generation, so it will be a no-op if there is no `-g...` option that enables debuginfo generation.

The [GCC manual](https://gcc.gnu.org/onlinedocs/gcc-12.2.0/gcc/Debugging-Options.html#Debugging-Options) is quite clear about this (emphasis mine):

> `-gsplit-dwarf`
> **If DWARF debugging information is enabled**, separate as much debugging information as possible into a separate output file with the extension .dwo. This option allows the build system to avoid linking files with debug information. To be useful, this option requires a debugger capable of reading .dwo files.

TL;DR: Always pass `-gsplit-dwarf` if supported, regardless of the build type, the compiler already handles this for us.

mcatanzaro replied:
Ah... good idea. :P

(Assuming this doesn't break distros, which we'll find out soon enough.)

aperezdc replied:
If that happened, we can change things so instead of _appending_ the flags, we would _prepend_ them, and distros can still override our default by setting `C[XX]FLAGS=-gno-split-dwarf` in the environment before invoking CMake. That will result in both flags in the command line, as in `-gsplit-dwarf -foo -bar -gno-split-dwarf` but the flag provided by the distro packager would take precedence.

mcatanzaro replied:
Um, good point: the rest of the code in this file is very careful to prepend rather than append almost always, and only ever append in cases where WebKit absolutely needs a certain flag (like `-fno-exceptions` or `-fno-rtti`). So the implementation here is backwards and should be flipped. Preexisting problem, though.

aperezdc commented on Source/cmake/OptionsCommon.cmake:
```diff
@@ -154,7 +154,7 @@ if (USE_THIN_ARCHIVES)
 endif ()
 
 set(ENABLE_DEBUG_FISSION_DEFAULT OFF)
-if (CMAKE_BUILD_TYPE STREQUAL "Debug")
```
Remove this line and its corresponding `endif()`, then dedent the contents of the block. That will always pass `-gsplit-dwarf` to the compiler flags regardless of the build type, which is a better option [as outlined here](https://github.com/WebKit/WebKit/pull/9602#discussion_r1098443968).

aperezdc reviewed (changes requested):
Commented with a saner and simpler approach 😉

calvaris committed:
[GTK][WPE] Enable debug fission also when -g is present in the flags
https://bugs.webkit.org/show_bug.cgi?id=251685

Reviewed by Adrian Perez de Castro.

* Source/cmake/OptionsCommon.cmake:

Canonical link: https://commits.webkit.org/259957@main
